Glide - comments

malinajirka opened the pull request:
Fixes #7962 

Replaces Volley with Glide in the `Comments` and `Reader - Comments` sections. 
Adds support for loading animated GIF images in both sections.

Adds new subclass of Html.ImageGetter (`WPCustomImageGetter`), which takes care of replacing `<img>` tags with actual images using Glide. Glide supports loading resources into ImageViews or so called Targets. Since we display comments into a TextView using Html.fromHtml() method, we must implement a custom Target (`WPRemoteResourceViewTarget`) with support for loading multiple images into a single TextView.  `WPRemoteResourceViewTarget` also supports displaying placeholder and replacing it when the request finishes with remote/error image. More info about Targets can be found [here](http://bumptech.github.io/glide/doc/targets.html) and [here](https://bumptech.github.io/glide/doc/resourcereuse.html#attempting-to-load-two-different-resources-into-the-same-target).

Since this is quite a significant change and it's easy to miss something, can you please both review it. Thanks!!

To test:
1. Post a comment with 2+ gif images (using web)
2. Go to My Site -> Comments
3. Make sure avatars look as expected
4. Open detail of the comment
5. Make sure avatar looks as expected
6. Notice loading placeholder is being displayed
7. Notice all images are animated
8. Click on the name of the blog in the header
9. Click on comments in the bottom bar
10. Notice all images are animated
![gif-comments2](https://user-images.githubusercontent.com/2261188/42751870-770a8c6a-88ec-11e8-9482-28c9f1e4e31e.gif)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/ImageManager.kt:
```diff
@@ -47,6 +50,14 @@ class ImageManager @Inject constructor(val placeholderManager: ImagePlaceholderM
                 .into(imageView)
     }
 
+    fun load(viewTarget: ViewTarget<TextView, Drawable>, imageType: ImageType, imgUrl: String) {
```
Just curious, is it at all possible to avoid this many different `load` methods in `ImageManager`? Can we use a builder pattern or named parameters with default values to avoid it? This, of course, doesn't need to be part of this PR, but just a general concern.

malinajirka replied:
I've consider it before starting this project, however I felt it'd be nice to have limited options how to load images, so we can easily be consistent across the app. I also didn't want to have any Glide classes in the interface, but it could be solved by wrapping GlideRequest, I guess. However, I don't have a strong opinion on this. Wdyt?

oguzkocer replied:
I think some kind of refactor would be great for this. I have to look into it more to figure out what the best way might be, but I think we simply have too many load functions right now.

As I said, this is not related to this PR, but something to consider before we finalize the Glide project.

malinajirka replied:
I'd appreciate any help with it. We might allocate 30 minutes and resolve it during a call. Wdys?

oguzkocer replied:
Sure thing. Do you want to open an issue for this and assign it to both of us, so we can check it out once all Glide PRs are merged in?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/GlideImageGetter.kt:
```diff
@@ -0,0 +1,74 @@
+import java.util.HashSet
+import javax.inject.Inject
+
+class GlideImageGetter(
```
Can we please add some comments about what this class is for? We don't have anything about it in the PR description either, so I don't know how to review this code. It'd be great if you can add why it's necessary and what it does, thanks!

oguzkocer replied:
Also, I am curious if we can find a better name for the class. I don't know what it does yet, but since it starts with `Glide` I just assumed it was part of the 3rd party library when I saw its usage.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/GlideRemoteResourceViewTarget.kt:
```diff
@@ -0,0 +1,116 @@
+
+import org.wordpress.android.util.R
+
+internal class GlideRemoteResourceViewTarget(
```
Similar to my comments for `GlideImageGetter`, can we add comments about what this class is about and maybe come up with a different name that doesn't start with `Glide`?

Also, is there a reason it's an internal class?

oguzkocer replied:
Because it's in the main module, I don't think it'll make any difference, so I didn't think there is a reason to break consistency here, but I don't feel strongly about it, I was just curious.

> I'm not a fan of having multiple classes declaration in a single file

For me, that depends on the classes. If that class is only used by one other class and shouldn't be used otherwise and it's short enough, I actually prefer to have it in the same file.

malinajirka replied:
I agree, it definitely depends and there are many valid cases.

malinajirka commented:
Thanks @oguzkocer ! I'm sorry for creating such a poor PR. I added some more info into the description and I added several comments into the code. Let me know if you have any questions!

malinajirka commented:
Hmm, I've just realized that calling `GlideApp.with(imageView.context)` instead of `GlideApp.with(activity/fragment)` in the ImageManager has some negative impact on performance. However, I'll address this in another PR (issue #8044).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/WPRemoteResourceViewTarget.kt:
```diff
@@ -0,0 +1,135 @@
+    private fun getScaledBounds(resource: Drawable, maxWidth: Int): Rect {
+        val imgWidth = resource.intrinsicWidth
+        val imgHeight = resource.intrinsicHeight
+        val xScale = imgWidth.toFloat() / maxWidth.toFloat()
```
To simplify the calculation below, we can simply do `Math.max(1.0f, imgWidth.toFloat() / maxWidth.toFloat())`.

malinajirka replied:
Good call;). I've just copied the code from WPImageGetter. I should have refactored it a bit.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/WPCustomImageGetter.kt:
```diff
@@ -0,0 +1,94 @@
+            targets.add(target)
+
+            return target.drawable
+        } ?: return null
```
Since you're already returning the whole statement, you don't need the `?: return null`, you can simply remove it.

malinajirka replied:
Ohh, cool. Thanks;)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/WPCustomImageGetter.kt:
```diff
@@ -0,0 +1,94 @@
+            imageManager.load(target, ImageType.UNKNOWN_DIMENSIONS, source)
+            targets.add(target)
+
+            return target.drawable
```
Since the statement is being returned, we don't need the `return` here, we can just say `target.drawable`.

malinajirka replied:
This is something I can't get use to. I guess I like explicit return statements. However, I've removed it since it's a common practice in Kotlin.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/image/ImageType.kt:
```diff
@@ -5,5 +5,6 @@ enum class ImageType {
     BLAVATAR,
-    THEME
+    THEME,
+    UNKNOWN_DIMENSIONS
```
I think we should just say `UNKNOWN` here. The enum is name as `ImageType` and unknown dimensions as a type doesn't make sense to me. We can do `UNKNOWN_TYPE` but I think that's clear from the enum name. What do you think?

malinajirka replied:
👍 I renamed it from UNKNOWN -> UNKNOWN_DIMENSION and I wasn't sure whether it was a good call or not. You made this decision easier -> I've renamed it back to UNKNOWN.

oguzkocer commented on libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/helpers/WPImageGetter.java:
```diff
@@ -20,7 +20,12 @@
  * adapted from existing ImageGetter code in NoteCommentFragment
+ *
+ *
+ * @deprecated Use org.wordpress.android.util.image.getters.WPCustomImageGetter instead.
```
It looks like there is only one other place we use this class, should we just update that as well, if not in this PR, maybe in the next one? Maybe you were already planning to do so, so ignore this comment if that's the case :)

malinajirka replied:
;) I was planning to update it in another PR and remove the WPImageGetter entirely.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/WPCustomImageGetter.kt:
```diff
@@ -0,0 +1,94 @@
+    override fun getDrawable(url: String): Drawable? {
+        var source = url
+
+        // images in reader comments may skip "http:" (no idea why) so make sure to add protocol here
```
I'm not sure how exactly do you propose to fix it there. This class receives HTML string and scans it for <img> tags, when it reaches one it calls `getDrawable(src/url)` and replaces the <img> tag with the actual image (ImageSpan).  We'd have to create another scanner/replacer and look for invalid urls in <img> tags.

oguzkocer replied:
We could preprocess the passed in html string, but it's not that important.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/util/image/getters/WPRemoteResourceViewTarget.kt:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    /**
+     * Drawable wrapper so we can replace placeholder with remote/error resource, after the requests finishes.
```
The idea is following
1. We need to synchronously return drawable in WPCustomImageGetter.getDrawable(...)
2. If we return direct drawable - let's say placeholder -> afaik we won't be able to replace it with the actual image, after it's fetched.
==> This wrapper just adds us ability to change the content of the drawable after the asynchronous call finishes.

As most of this code, it's heavily based on [current implementation of Html.ImageGetter](https://github.com/wordpress-mobile/WordPress-Android/blob/80d23e2c99cb8120fb1ba80cecc4b43bbb06eb4a/libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/helpers/WPImageGetter.java#L119). I didn't think of any better solution:(. If you have any suggestion, I'll be more than happy to refactor this.

malinajirka commented:
Thanks @oguzkocer! No worries at all, I'll be happy to answer any of your questions. I've fixed all the issues you mentioned in comments.

>I guess the thing that confuses me the most is the fact that we are inheriting view components in both of these new classes.

I'm not sure I understand what you mean by that. None of the classes inherits from View class (if you mean android.view.View).

`WPCustomImageGetter` is a class which replaces `img` tag with `ImageSpan` and loads an image into it. 
It also contains logic for clearing any previous targets/requests from the TextView - it's not ideal, on the other hand no-one else but this class knows about the targets. Moving this to a separate class seems like an overkill - however if you prefer it, I don't mind moving it. 
I've moved the implementation of `Drawable.Callback` to a separate class as this seemed really unrelated.

`WPRemoteResourceViewTarget` is a Glide specific class (that's why I used `Glide` in it's name). I don't feel like it contains any logic which should/could be extracted, except of the `getScaledBounds` which I moved to `ImageUtils` class.

I'll be happy to separate the concerns even more. I 100% agree that View classes shouldn't contain any logic. But I don't consider either of the classes as View classes.

>Also, I appreciate you adding comments to the classes, but I am afraid it might not be enough for a newcomer who is unfamiliar with this part of the code. Again, that might be just me 🤷‍♂️ Sorry about that!

Anything in particular you are missing? I think newcomer should always read javadoc of parent class. I'm not sure what else should I add:(. Nevertheless I'll be also happy to improve it if you have any specific suggestions.

theck13 commented:
@oguzkocer, your review was good enough for me.  You can merge it whenever you're ready.
